MovingBlocks/Terasology
# Convert the module build script to a precompiled plugin

### Contains

This moves `templates/build.gradle` to gradle's `buildSrc/` directory where it becomes a [precompiled script plugin](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins).

Also, it is now a Kotlin gradle script instead of a Groovy gradle script.


### How to test

* Have a workspace with some module sources.
* `groovyw module refresh` to replace their `build.gradle` files.
* make sure modules build and run as usual.
* test the _publishing_ actions for the modules, the stuff configured by config/gradle/publish.gradle.  
  I'm not sure how to test that part and it's the most likely to have broken.


### Discussion

This introduces some new tech for the build process so I've tried to keep the scope of this small. The things under consideration here are:

* a module's build.gradle becomes as short as `plugins { id("terasology-module") }`
* starting to migrate the gradle configuration to Kotlin

I tried to transliterate the groovy config to kotlin *as directly as possible.* There's little here that cleans things up, and it's not a showcase of the best way to write something in Kotlin. I figured _changing languages_ is a big enough change without other reorganization, refactoring, or redesign on top.

GitHub's diff for the PR just shows a big delete in one file and add in another, but you can see more detail in this commit: https://github.com/MovingBlocks/Terasology/pull/3993/commits/0a352a72c887e16077f95b91c6f56937b06c4b90#diff-873dc4d71db70d570c73ef0cb16a8731

It has already demonstrated some practical benefits:

* Because the only thing the gradle file in the module subdirectory does is apply the plugin, there's no longer need to `groovyw module refresh` to update module build configs. Changing it in one place in buildSrc is enough.
* Because the script plugin is precompiled, the configuration phase of big multi-project workspaces is _a lot faster._ It only has to compile that mess once after it changes, not a hundred times over for the copies of it in a hundred module directories.
